boards: cy8proto_041tp: Enable GPIO driver and update clock driver

dharunkrithikk opened the pull request:
Add initial support for Infineon PSoC4 GPIO and CLOCK driver updates.

dharunkrithikk committed:
dt-bindings: clock: add Infineon PSoC4 clock source bindings

Add clock source definitions for the Infineon PSoC4 series.
This introduces the \`ifx_clock_source_psoc4xx.h\` header with
HFCLK divider macros and integrates it into the board clock
source selection.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
drivers: clock_control: update infineon clock drivers

Update the Infineon clock control drivers to support PSoC4 and
enhance error handling.

- Add support for ILO, WCO, and EXT clocks in fixed clock driver.
- Update fixed factor clock driver to handle PSoC4 specific HF
  clock source setting and add error checking.
- Update peripheral clock driver to support fractional dividers
- Improve error handling across all updated drivers.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
boards: cy8cproto_041tp: refactor dts and enable UART

Refactor the CY8CPROTO-041TP board definition by extracting common
configurations into \`cy8cproto_041tp_common.dtsi\`. This improves
reusability and structure.

Additionally, refactored the nodes accordingly and configure necessary
system clocks:
- Configure 16-bit peripheral dividers.
- Enable clock paths.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

manoj-aerlync committed:
boards: cy8cproto_041tp: Add dts for button and led

- Add LED0 and SW0 support to CY8CPROTO-041TP board
- This enables applications to use led0 and sw0 through
  standard Zephyr subsystems without additional board-specific
  configuration.

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

manoj-aerlync committed:
drivers: gpio: Add Infineon PSoC4 GPIO driver

Introduces a GPIO driver that provides generic interfaces
for both interrupt-driven inputs and controllable outputs.

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk commented:
Could you please review this PR @billwatersiii @henrikbrixandersen @jsbatch @mcatee-infineon @sreeramIfx @teburd

teburd commented on boards/infineon/cy8cproto_041tp/cy8cproto_041tp.dts:
```diff
@@ -21,4 +21,29 @@
+	gpio_keys {
+		compatible = "gpio-keys";
+
+		user_bt: button_0 {
```
user_btn

dharunkrithikk commented on drivers/clock_control/clock_control_infineon_peri_clock.c:
```diff
@@ -31,10 +31,26 @@ struct ifx_peri_clock_data {
+
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+
+#define _IFX_PSOC4_PCLK_TCPWM_CLOCKS0 PCLK_TCPWM_CLOCKS0
```
Good point. I see how this macro is SOC-specific. Since #94679 is still [DNM] (Do Not Merge), should we keep this temporary workaround as-is for now, or is there a better pattern we should follow meantime ?

teburd replied:
Yes it’s fine for now

dharunkrithikk commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Cypress Semiconductor Corporation (an Infineon company) or
+ * Copyright (c) 2022-2025 Cypress Semiconductor Corporation (an Infineon company) or
```
changed

dharunkrithikk commented on include/zephyr/dt-bindings/clock/ifx_clock_source_psoc4xx.h:
```diff
@@ -0,0 +1,24 @@
+#define IFX_PATH_PSOC4_PUMP_IMO   1
+#define IFX_PATH_PSOC4_PUMP_HFCLK 2
+
+#define IFX_PSOC4_HFCLK_DIV(div)                                                                   \
```
Thank you for pointing out, must have missed while rebasing,
 this macro is used in clock_control_infineon_fixed_factor_clock.c have added now
you can see in ifx_clock_source_common.h uses 
```
#define IFX_CLK_HF_NO_DIVIDE    0  /**< don't divide clkHf */
#define IFX_CLK_HF_DIVIDE_BY_2  1  /**< divide clkHf by 2 */ 
#define IFX_CLK_HF_DIVIDE_BY_3  2  /**< divide clkHf by 3 */
#define IFX_CLK_HF_DIVIDE_BY_4  3  /**< divide clkHf by 4 */
#define IFX_CLK_HF_DIVIDE_BY_5  4  /**< divide clkHf by 5 */
#define IFX_CLK_HF_DIVIDE_BY_6  5  /**< divide clkHf by 6 */
#define IFX_CLK_HF_DIVIDE_BY_7  6  /**< divide clkHf by 7 */ 
#define IFX_CLK_HF_DIVIDE_BY_8  7  /**< divide clkHf by 8 */
...
```
whereas in bare metal uses clk divider like this, hence to use the common.h macro by user 
to be generic in nature for all soc, this macro is used and specific for psoc4
```
typedef enum                                      
{                         
    CY_SYSCLK_NO_DIV = 0U,    /**< don't divide */
    CY_SYSCLK_DIV_2  = 1U,    /**< divide by 2 */ 
    CY_SYSCLK_DIV_4  = 2U,    /**< divide by 4 */ 
    CY_SYSCLK_DIV_8  = 3U     /**< divide by 8 */ 
} cy_en_sysclk_dividers_t;                
```

dharunkrithikk commented on dts/arm/infineon/psoc4/psoc4100tp/system_clocks.dtsi:
```diff
@@ -5,11 +5,12 @@
-#define DIV_16_BIT   1
-#define DIV_16_5_BIT 2
-#define DIV_24_5_BIT 3
+#define DIV_16_BIT	1
```
reverted back

teburd commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +196,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
-	uint32_t pins = GPIO_PRT_INTR_MASKED(base);
+	uint32_t pending;
+
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
```
non-blocking, but we should revisit this ifdefry

JarmouniA commented on include/zephyr/dt-bindings/clock/ifx_clock_source_psoc4xx.h:
```diff
@@ -0,0 +1,24 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Please add some description here of what these Macros are and what DT compatibles/properties they are supposed to be used with.

JarmouniA commented on dts/arm/infineon/psoc4/psoc4100tp/system_clocks.dtsi:
```diff
@@ -32,63 +33,49 @@
-		peri0_group0_16bit_0: peri0_group0_16bit_0 {
+	/*
+	 * Note: peri-group property is not required for psoc4100t but to satisfy the
+	 * clk driver it is initialized to 0
```
That should be fixed in the driver or the binding, instead of having this workaround in DT.

dharunkrithikk replied:
changed

JarmouniA commented on dts/arm/infineon/psoc4/psoc4100tp/psoc4100tp.dtsi:
```diff
@@ -28,13 +27,13 @@
 
 		gpio_prt0: gpio@40040000 {
-			compatible = "infineon,cat1-gpio";
+			compatible = "infineon,gpio";
```
The compatibles renaming should be in a separate 1st commit.

teburd replied:
The binding only exists for infineon,gpio so this does need to be renamed. A commit is fine here to do that, but as @JarmouniA it should be a commit in the PR that renames these.

JarmouniA commented on drivers/clock_control/Kconfig.infineon:
```diff
@@ -43,4 +43,5 @@ config CLOCK_CONTROL_IFX_FIXED_FACTOR_CLOCK_INIT_PRIORITY
 	  Infineon fixed factor clock control driver initialization priority.
 	  This should be lower than the priority of fixed factor clock driver
 	  to ensure the clock paths are initialized before.
+
```
Please do not mix formatting changes with functional ones in the same commit, and avoid them unless they are necessary.

JarmouniA commented on boards/infineon/cy8cproto_041tp/cy8cproto_041tp.dts:
```diff
@@ -21,6 +21,31 @@
+		compatible = "gpio-leds";
+		status = "okay";
+
+		led0: led_0 {
```
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#generic-names-recommendation

dharunkrithikk commented on dts/arm/infineon/psoc4/psoc4100tp/system_clocks.dtsi:
```diff
@@ -32,63 +33,49 @@
 			#clock-cells = <0>;
 			compatible = "infineon,fixed-clock";
-			clock-frequency = <32000>; /* 32 kHz */
+			clock-frequency = <0>;
```
@JarmouniA please refer this comment, so i have moved this to board.dts https://github.com/zephyrproject-rtos/zephyr/pull/100894#discussion_r2614168213

dharunkrithikk commented on dts/arm/infineon/psoc4/psoc4100tp/system_clocks.dtsi:
```diff
@@ -32,63 +33,49 @@
 			#clock-cells = <0>;
 			compatible = "infineon,fixed-clock";
-			clock-frequency = <48000000>; /* 48 MHz external input */
+			clock-frequency = <48000000>; /* 48 MHz */
```
changed

dharunkrithikk committed:
dts: infineon: update PSoC4100TP SoC nodes

Update all Infineon PSoC4100TP device tree nodes to use current
binding naming convention by removing the 'cat1-' prefix

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
dts: infineon: update PSoC4100TP system clock nodes

Update the system clock device tree definitions for the Infineon
PSoC4100TP series.

Refactor the peripheral divider nodes in `system_clocks.dtsi` to
align with the hardware specifications. This includes renaming nodes
and updating group associations for correct clock configuration.
Update `infineon,peri-div.yaml` binding to support the changes.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
